In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("hw7.ipynb")

# CPSC 330 - Applied Machine Learning 

## Homework 7: Word embeddings and topic modeling 
**Due date: See the [Calendar](https://htmlpreview.github.io/?https://github.com/UBC-CS/cpsc330/blob/master/docs/calendar.html).**

## Imports

In [2]:
import os

%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline, make_pipeline

<br><br>

<!-- BEGIN QUESTION -->

<div class="alert alert-info">
    
## Submission instructions
<hr>
rubric={points}

**Please be aware that this homework assignment requires installation of several packages in your course environment. It's possible that you'll encounter installation challenges, which might be frustrating. However, remember that solving these issues is not wasting time but it is an essential skill for anyone aspiring to work in data science or machine learning.**

Follow the [homework submission instructions](https://github.com/UBC-CS/cpsc330-2024W1/blob/main/docs/homework_instructions.md). 

**You may work in a group on this homework and submit your assignment as a group.** Below are some instructions on working as a group.  
- The maximum group size is 2. 
- Use group work as an opportunity to collaborate and learn new things from each other. 
- Be respectful to each other and make sure you understand all the concepts in the assignment well. 
- It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline. 
- You can find the instructions on how to do group submission on Gradescope [here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members).


When you are ready to submit your assignment do the following:

1. Run all cells in your notebook to make sure there are no errors by doing `Kernel -> Restart Kernel and Clear All Outputs` and then `Run -> Run All Cells`. 
2. Notebooks with cell execution numbers out of order or not starting from “1” will have marks deducted. Notebooks without the output displayed may not be graded at all (because we need to see the output in order to grade your work).
3. Upload the assignment using Gradescope's drag and drop tool. Check out this [Gradescope Student Guide](https://lthub.ubc.ca/guides/gradescope-student-guide/) if you need help with Gradescope submission.
4. Make sure that the plots and output are rendered properly in your submitted file. 
5. If the .ipynb file is too big and doesn't render on Gradescope, also upload a pdf or html in addition to the .ipynb. If the pdf or html also fail to render on Gradescope, please create two files for your homework: hw6a.ipynb with Exercise 1 and hw6b.ipynb with Exercises 2 and 3 and submit these two files in your submission.  
</div>

_Points:_ 2

<!-- END QUESTION -->

<br><br><br><br>

## Exercise 1:  Exploring pre-trained word embeddings <a name="1"></a>
<hr>

In lecture 18, we talked about natural language processing (NLP). Using pre-trained word embeddings is very common in NLP. It has been shown that pre-trained word embeddings work well on a variety of text classification tasks. These embeddings are created by training a model like Word2Vec on a huge corpus of text such as a dump of Wikipedia or a dump of the web crawl. 

A number of pre-trained word embeddings are available out there. Some popular ones are: 

- [GloVe](https://nlp.stanford.edu/projects/glove/)
    * trained using [the GloVe algorithm](https://nlp.stanford.edu/pubs/glove.pdf) 
    * published by Stanford University 
- [fastText pre-trained embeddings for 294 languages](https://fasttext.cc/docs/en/pretrained-vectors.html) 
    * trained using the fastText algorithm
    * published by Facebook
    
In this exercise, you will be exploring GloVe Wikipedia pre-trained embeddings. The code below loads the word vectors trained on Wikipedia using an algorithm called Glove. You'll need `gensim` package in your cpsc330 conda environment to run the code below. 

```
> conda activate cpsc330
> conda install -c anaconda gensim
```

In [5]:
import gensim
import gensim.downloader

print(list(gensim.downloader.info()["models"].keys()))

['fasttext-wiki-news-subwords-300', 'conceptnet-numberbatch-17-06-300', 'word2vec-ruscorpora-300', 'word2vec-google-news-300', 'glove-wiki-gigaword-50', 'glove-wiki-gigaword-100', 'glove-wiki-gigaword-200', 'glove-wiki-gigaword-300', 'glove-twitter-25', 'glove-twitter-50', 'glove-twitter-100', 'glove-twitter-200', '__testing_word2vec-matrix-synopsis']


In [6]:
# This will take a while to run when you run it for the first time.
import gensim.downloader as api

glove_wiki_vectors = api.load("glove-wiki-gigaword-100")

[==================================================] 100.0% 128.1/128.1MB downloaded


In [7]:
len(glove_wiki_vectors)

400000

There are 400,000 word vectors in this pre-trained model. 

Now that we have GloVe Wiki vectors loaded in `glove_wiki_vectors`, let's explore the embeddings. 

<br><br>

<!-- BEGIN QUESTION -->

### 1.1 Word similarity using pre-trained embeddings
rubric={points}

**Your tasks:**

- Come up with a list of 4 words of your choice and find similar words to these words using `glove_wiki_vectors` embeddings.

<div class="alert alert-warning">

Solution_1.1
    
</div>

_Points:_ 2

In [11]:
similar_words = glove_wiki_vectors.most_similar('badminton')
print("Similar words to 'badminton':")
for word, score in similar_words:
    print(f"  {word}: {score:.4f}")

Similar words to 'badminton':
  volleyball: 0.8076
  tennis: 0.7742
  gymnastics: 0.7268
  weightlifting: 0.7062
  snooker: 0.7043
  taekwondo: 0.6897
  championships: 0.6888
  judo: 0.6794
  softball: 0.6760
  netball: 0.6698


In [12]:
similar_words = glove_wiki_vectors.most_similar('baseball')
print("Similar words to 'baseball':")
for word, score in similar_words:
    print(f"  {word}: {score:.4f}")

Similar words to 'baseball':
  basketball: 0.7645
  leagues: 0.7626
  football: 0.7606
  league: 0.7323
  hockey: 0.7312
  sports: 0.7067
  nfl: 0.7048
  soccer: 0.7028
  mlb: 0.6976
  yankees: 0.6953


In [17]:
similar_words = glove_wiki_vectors.most_similar('diving')
print("Similar words to 'diving':")
for word, score in similar_words:
    print(f"  {word}: {score:.4f}")

Similar words to 'diving':
  dive: 0.7657
  swimming: 0.7169
  scuba: 0.7153
  snorkeling: 0.6957
  diver: 0.6329
  swim: 0.6257
  snorkelling: 0.6025
  divers: 0.5953
  surfing: 0.5818
  skiing: 0.5737


In [19]:
similar_words = glove_wiki_vectors.most_similar('hockey')
print("Similar words to 'hockey':")
for word, score in similar_words:
    print(f"  {word}: {score:.4f}")

Similar words to 'hockey':
  basketball: 0.8042
  football: 0.7834
  nhl: 0.7604
  soccer: 0.7441
  baseball: 0.7312
  league: 0.7092
  skating: 0.6704
  lacrosse: 0.6692
  team: 0.6620
  games: 0.6572


<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.2 Word similarity using pre-trained embeddings
rubric={points}

**Your tasks:**

1. Calculate cosine similarity for the following word pairs (`word_pairs`) using the [`similarity`](https://radimrehurek.com/gensim/models/keyedvectors.html?highlight=similarity#gensim.models.keyedvectors.KeyedVectors.similarity) method of `glove_wiki_vectors`.

In [20]:
word_pairs = [
    ("coast", "shore"),
    ("clothes", "closet"),
    ("old", "new"),
    ("smart", "intelligent"),
    ("dog", "cat"),
    ("tree", "lawyer"),
]

<div class="alert alert-warning">

Solution_1.2
    
</div>

_Points:_ 2

In [21]:
for word1, word2 in word_pairs:
    try:
        similarity = glove_wiki_vectors.similarity(word1, word2)
        print(f"Cosine similarity between '{word1}' and '{word2}': {similarity:.4f}")
    except KeyError as e:
        print(f"Warning: One of the words '{word1}' or '{word2}' not found in vocabulary")

Cosine similarity between 'coast' and 'shore': 0.7000
Cosine similarity between 'clothes' and 'closet': 0.5463
Cosine similarity between 'old' and 'new': 0.6432
Cosine similarity between 'smart' and 'intelligent': 0.7553
Cosine similarity between 'dog' and 'cat': 0.8798
Cosine similarity between 'tree' and 'lawyer': 0.0767


<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.3 Representation of all words in English
rubric={points}

**Your tasks:**

1. The vocabulary size of Wikipedia embeddings is quite large. The `test_words` list below contains a few new words (called neologisms) and biomedical domain-specific abbreviations. Write code to check whether `glove_wiki_vectors` have representation for these words or not. 
> If a given word `word` is in the vocabulary, `word in glove_wiki_vectors` will return True. 

In [22]:
test_words = [
    "covididiot",
    "fomo",
    "frenemies",
    "anthropause",
    "photobomb",
    "selfie",
    "pxg",  # Abbreviation for pseudoexfoliative glaucoma
    "pacg",  # Abbreviation for primary angle closure glaucoma
    "cct",  # Abbreviation for central corneal thickness
    "escc",  # Abbreviation for esophageal squamous cell carcinoma
]

<div class="alert alert-warning">

Solution_1_3
    
</div>

_Points:_ 2

In [23]:
for word in test_words:
    exists = word in glove_wiki_vectors
    status = "exists" if exists else "does not exist"
    print(f"'{word}' {status} in the vocabulary")

'covididiot' does not exist in the vocabulary
'fomo' does not exist in the vocabulary
'frenemies' exists in the vocabulary
'anthropause' does not exist in the vocabulary
'photobomb' does not exist in the vocabulary
'selfie' does not exist in the vocabulary
'pxg' does not exist in the vocabulary
'pacg' does not exist in the vocabulary
'cct' exists in the vocabulary
'escc' exists in the vocabulary


<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.4 Stereotypes and biases in embeddings
rubric={points}

Word vectors contain lots of useful information. But they also contain stereotypes and biases of the texts they were trained on. In the lecture, we saw an example of gender bias in Google News word embeddings. Here we are using pre-trained embeddings trained on Wikipedia data. 

**Your tasks:**

1. Explore whether there are any worrisome biases or stereotypes present in these embeddings by trying out at least 4 examples. You can use the following two methods or other methods of your choice to explore this. 
    - the `analogy` function below which gives word analogies (an example shown below)
    - [similarity](https://radimrehurek.com/gensim/models/keyedvectors.html?highlight=similarity#gensim.models.keyedvectors.KeyedVectors.similarity) or [distance](https://radimrehurek.com/gensim/models/keyedvectors.html?highlight=distance#gensim.models.keyedvectors.KeyedVectors.distances) methods (an example is shown below)

> Note that most of the recent embeddings are de-biased. But you might still observe some biases in them. Also, not all stereotypes present in pre-trained embeddings are necessarily bad. But you should be aware of them when you use them in your models. 

In [24]:
def analogy(word1, word2, word3, model=glove_wiki_vectors):
    """
    Returns analogy word using the given model.

    Parameters
    --------------
    word1 : (str)
        word1 in the analogy relation
    word2 : (str)
        word2 in the analogy relation
    word3 : (str)
        word3 in the analogy relation
    model :
        word embedding model

    Returns
    ---------------
        pd.dataframe
    """
    print("%s : %s :: %s : ?" % (word1, word2, word3))
    sim_words = model.most_similar(positive=[word3, word2], negative=[word1])
    return pd.DataFrame(sim_words, columns=["Analogy word", "Score"])

Examples of using analogy to explore biases and stereotypes.  

In [25]:
analogy("man", "doctor", "woman")

man : doctor :: woman : ?


Analogy word     Score
0        nurse  0.773523
1    physician  0.718943
2      doctors  0.682433
3      patient  0.675068
4      dentist  0.672603
5     pregnant  0.664246
6      medical  0.652045
7      nursing  0.645348
8       mother  0.639333
9     hospital  0.638750

In [26]:
glove_wiki_vectors.similarity("aboriginal", "success")

0.1428324

In [27]:
glove_wiki_vectors.similarity("white", "success")

0.351824

<div class="alert alert-warning">

Solution_1_4
    
</div>

_Points:_ 4

In [28]:
#Gender Bias in professions
print(f"Similarity between 'nurse' and 'female':", glove_wiki_vectors.similarity("nurse", "female"))
print(f"Similarity between 'nurse' and 'male':", glove_wiki_vectors.similarity("nurse", "male"))b

man : doctor :: woman : ?
Similarity between 'nurse' and 'female': 0.46290737
Similarity between 'nurse' and 'male': 0.39995456


In [29]:
# Example 2: Cultural and ethnic stereotypes
words_to_check = ["educated", "poverty", "crime", "success"]
ethnicities = ["asian", "african", "european", "hispanic"]

for ethnicity in ethnicities:
    print(f"\nSimilarities for '{ethnicity}':")
    for word in words_to_check:
        similarity = glove_wiki_vectors.similarity(ethnicity, word)
        print(f"- with '{word}': {similarity:.4f}")


Similarities for 'asian':
- with 'educated': 0.2095
- with 'poverty': 0.4143
- with 'crime': 0.2857
- with 'success': 0.4309

Similarities for 'african':
- with 'educated': 0.3677
- with 'poverty': 0.5122
- with 'crime': 0.3481
- with 'success': 0.3853

Similarities for 'european':
- with 'educated': 0.2519
- with 'poverty': 0.2904
- with 'crime': 0.3276
- with 'success': 0.5152

Similarities for 'hispanic':
- with 'educated': 0.3824
- with 'poverty': 0.3769
- with 'crime': 0.3502
- with 'success': 0.2897


In [32]:
# Example 3: Age-related bias
age_terms = ["old", "young"]
attributes = ["competent", "rich", "productive"]

for age in age_terms:
    print(f"\nSimilarities for '{age}':")
    for attr in attributes:
        similarity = glove_wiki_vectors.similarity(age, attr)
        print(f"- with '{attr}': {similarity:.4f}")


Similarities for 'old':
- with 'competent': 0.1696
- with 'rich': 0.4015
- with 'productive': 0.2191

Similarities for 'young':
- with 'competent': 0.3665
- with 'rich': 0.5292
- with 'productive': 0.3582


In [36]:
# Example 4: Education-related bias
edu_terms = ["educated", "uneducated", "university", "dropout"]
attributes = ["successful", "intelligent", "wealthy", "criminal"]

for term in edu_terms:
    print(f"\nSimilarities for '{term}':")
    for attr in attributes:
            similarity = glove_wiki_vectors.similarity(term, attr)
            print(f"- with '{attr}': {similarity:.4f}")


Similarities for 'educated':
- with 'successful': 0.3909
- with 'intelligent': 0.4535
- with 'wealthy': 0.5864
- with 'criminal': 0.1732

Similarities for 'uneducated':
- with 'successful': 0.0247
- with 'intelligent': 0.4124
- with 'wealthy': 0.3658
- with 'criminal': 0.0363

Similarities for 'university':
- with 'successful': 0.3606
- with 'intelligent': 0.1985
- with 'wealthy': 0.2369
- with 'criminal': 0.3150

Similarities for 'dropout':
- with 'successful': 0.0670
- with 'intelligent': 0.0237
- with 'wealthy': 0.1315
- with 'criminal': 0.0834


<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.5 Discussion
rubric={points}

**Your tasks:**
1. Discuss your observations from 1.4. Are there any worrisome biases in these embeddings trained on Wikipedia?   
2. Give an example of how using embeddings with biases could cause harm in the real world.

<div class="alert alert-warning">

Solution_1_5
    
</div>

_Points:_ 4

1. 
1)When researching on gender stereotype on professions, there is higher similarity between 'nurse' and 'female' compares to 'male', suggesting a stereotype to assume a nurse as a woman.

2)The Cultural and ethnic stereotypes analysis shows concerning ethnic biases. For example, "european" has the strongest association with "success" (0.5152) while having the lowest association with "poverty" (0.2904), whereas "african" shows the opposite pattern with the highest association with "poverty" (0.5122) and lower association with "success" (0.3853), reflecting harmful stereotypes present in the training data.

3)The Age-related analysis shows clear age-based bias, with "young" having consistently higher similarity scores for positive attributes like "competent" (0.3665), "rich" (0.5292), and "productive" (0.3582) compared to "old", which shows notably lower associations with these same qualities, particularly for competence (0.1696) and productivity (0.2191).

4)The education-related analysis shows strong educational bias, where being "educated" has high positive correlations with "wealthy" (0.5864) and "intelligent" (0.4535), while "dropout" has very low associations with these same positive attributes and the lowest scores across all positive traits. This bias is particularly stark in terms of success, where "educated" scores 0.3909 for "successful" while "dropout" scores only 0.0670, reflecting societal prejudices about educational status.

2. 
An automated hiring system using these biased word embeddings could severely discriminate against job candidates in multiple ways. The system would unfairly prefer young, white, educated male candidates while putting others at a disadvantage - like giving lower scores to older workers, people of colors, women, or those without college degrees. When someone faces several of these biases at once - like an older woman of color without a college degree - the negative impact would be even worse, making it much harder for them to get hired despite their actual skills and experience.

<!-- END QUESTION -->

<br><br><br><br>

## Exercise 2: Topic modeling 

The goal of topic modeling is discovering high-level themes in a large collection of texts. 

In this homework, you will explore topics in [the 20 newsgroups text dataset](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_20newsgroups.html) using `scikit-learn`'s `LatentDirichletAllocation` (LDA) model. 

Usually, topic modeling is used for discovering abstract "topics" that occur in a collection of documents when you do not know the actual topics present in the documents. But 20 newsgroups text dataset is labeled with categories (e.g., sports, hardware, religion), and you will be able to cross-check the topics discovered by your model with these available topics. 

The starter code below loads the train and test portion of the data and convert the train portion into a pandas DataFrame. For speed, we will only consider documents with the following 8 categories. 

In [37]:
from sklearn.datasets import fetch_20newsgroups

In [38]:
cats = [
    "rec.sport.hockey",
    "rec.sport.baseball",
    "soc.religion.christian",
    "alt.atheism",
    "comp.graphics",
    "comp.windows.x",
    "talk.politics.mideast",
    "talk.politics.guns",
]  # We'll only consider these categories out of 20 categories for speed.

newsgroups_train = fetch_20newsgroups(
    subset="train", remove=("headers", "footers", "quotes"), categories=cats
)
X_news_train, y_news_train = newsgroups_train.data, newsgroups_train.target
df = pd.DataFrame(X_news_train, columns=["text"])
df["target"] = y_news_train
df["target_name"] = [
    newsgroups_train.target_names[target] for target in newsgroups_train.target
]
df

text  target  \
0     You know, I was reading 18 U.S.C. 922 and some...       6   
1     \n\n\nIt's not a bad question: I don't have an...       1   
2     \nActuallay I don't, but on the other hand I d...       1   
3     The following problem is really bugging me,\na...       2   
4     \n\n  This is the latest from UPI \n\n     For...       7   
...                                                 ...     ...   
4558  Hi Everyone ::\n\nI am  looking for  some soft...       1   
4559  Archive-name: x-faq/part3\nLast-modified: 1993...       2   
4560  \nThat's nice, but it doesn't answer the quest...       6   
4561  Hi,\n     I just got myself a Gateway 4DX-33V ...       2   
4562  \n\n[h] \tThe Armenians in Nagarno-Karabagh ar...       7   

                target_name  
0        talk.politics.guns  
1             comp.graphics  
2             comp.graphics  
3            comp.windows.x  
4     talk.politics.mideast  
...                     ...  
4558          comp.graphics  
4559         comp.windows.x  
4560     talk.politics.guns  
4561         comp.windows.x  
4562  talk.politics.mideast  

[4563 rows x 3 columns]

In [39]:
newsgroups_train.target_names

['alt.atheism',
 'comp.graphics',
 'comp.windows.x',
 'rec.sport.baseball',
 'rec.sport.hockey',
 'soc.religion.christian',
 'talk.politics.guns',
 'talk.politics.mideast']

<br><br>

<!-- BEGIN QUESTION -->

### 2.1 Preprocessing using [spaCy](https://spacy.io/)
rubric={points}

Preprocessing is a crucial step before carrying out topic modeling and it markedly affects topic modeling results. In this exercise, you'll prepare the data using [spaCy](https://spacy.io/) for topic modeling. 

**Your tasks:** 

- Write code using [spaCy](https://spacy.io/) to preprocess the `text` column in the given dataframe `df` and save the processed text in a new column called `text_pp` within the same dataframe.

If you do not have spaCy in your course environment, you'll have to [install it](https://spacy.io/usage) and download the pretrained model en_core_web_md. 

`python -m spacy download en_core_web_md`


Note that there is no such thing as "perfect" preprocessing. You'll have to make your own judgments and decisions on which tokens are likely to be more informative for the given task. Some common text preprocessing steps for topic modeling include: 
- getting rid of slashes, new-line characters, or any other non-informative characters
- sentence segmentation and tokenization      
- replacing urls, email addresses, or numbers with generic tokens such as "URL",  "EMAIL", "NUM". 
- getting rid of other fairly unique tokens which are not going to help us in topic modeling  
- excluding stopwords and punctuation 
- lemmatization


> Check out [these available attributes](https://spacy.io/api/token#attributes) for `token` in spaCy which might help you with preprocessing. 

> You can also get rid of words with specific POS tags. [Here](https://universaldependencies.org/u/pos/) is the list of part-of-speech tags used in spaCy. 

> You may have to use regex to clean text before passing it to spaCy. Also, you might have to go back and forth between preprocessing in this exercise and and topic modeling in Exercise 2 before finalizing preprocessing steps. 

> Note that preprocessing the corpus might take some time. So here are a couple of suggestions: 1) During the debugging phase, work on a smaller subset of the data. 2) Once you finalize the preprocessing part, you might want to save the preprocessed data in a CSV and work with this CSV so that you don't run the preprocessing part every time you run the notebook. 
 


In [41]:
import spacy
nlp = spacy.load("en_core_web_md", disable=["parser", "ner"])

<div class="alert alert-warning">

Solution_2_1
    
</div>

_Points:_ 8

In [43]:
import re
def preprocess_text(text):
    # Step 1: Remove unwanted characters
    text = re.sub(r"\s+", " ", text)  
    text = re.sub(r"\n", " ", text)    
    text = re.sub(r"http\S+", "URL", text)  
    text = re.sub(r"\b\d+\b", "NUM", text)
    
    # Step 2: Process the text with spaCy (tokenization and lemmatization)
    doc = nlp(text)
    
    tokens = []
    for token in doc:
        if not token.is_stop and not token.is_punct and not token.like_email:
            if token.pos_ not in ["PRON", "DET", "PART", "SYM", "SPACE"]:  
                lemma = token.lemma_.lower().strip() 
                tokens.append(lemma)
    
    return " ".join(tokens)

df["text_pp"] = df["text"].apply(preprocess_text)

In [44]:
print(df[["text", "text_pp"]].head())

                                                text  \
0  You know, I was reading 18 U.S.C. 922 and some...   
1  \n\n\nIt's not a bad question: I don't have an...   
2  \nActuallay I don't, but on the other hand I d...   
3  The following problem is really bugging me,\na...   
4  \n\n  This is the latest from UPI \n\n     For...   

                                             text_pp  
0  know read num u.s.c. num sence wonder help u.s...  
1  bad question ref list algorithm think bit hard...  
2  actuallay hand support idea have newsgroup asp...  
3  follow problem bug appreciate help create wind...  
4  late upi foreign ministry spokesman ferhat ata...  


In [45]:
df.to_csv("preprocessed_20newsgroups.csv", index=False)

In [46]:
df.iloc[2:6]

text  target  \
2  \nActuallay I don't, but on the other hand I d...       1   
3  The following problem is really bugging me,\na...       2   
4  \n\n  This is the latest from UPI \n\n     For...       7   
5  Hi,\n   I'd like to subscribe to Leadership Ma...       5   

              target_name                                            text_pp  
2           comp.graphics  actuallay hand support idea have newsgroup asp...  
3          comp.windows.x  follow problem bug appreciate help create wind...  
4   talk.politics.mideast  late upi foreign ministry spokesman ferhat ata...  
5  soc.religion.christian  hi like subscribe leadership magazine wonder d...

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.2 Justification
rubric={points}

**Your tasks:**

- Outline the preprocessing steps you carried out in the previous exercise (bullet point format is fine), providing a brief justification when necessary. 

> You might want to wait to answer this question till you are done with Exercise 2 and you have finalized the preprocessing steps in 2.1. 

<div class="alert alert-warning">

Solution_2_2
    
</div>

_Points:_ 2

1. **Removal of Extra Whitespace and Newline Characters**:
   - Used regular expressions to remove unnecessary whitespace (`\s+`) and newline characters (`\n`). This cleans the text and standardizes the format for further processing.

2. **Replacement of URLs and Numbers with Generic Tokens**:
   - Replaced URLs with "URL" and numbers with "NUM" to remove any specific, irrelevant content that may not contribute meaningfully to the topics.

3. **Tokenization and Lemmatization**:
   - Utilized spaCy to split the text into individual tokens and lemmatize them to reduce words to their base form. Lemmatization helps in grouping similar words and reducing redundancy in the vocabulary.

4. **Exclusion of Stop Words, Punctuation, and Non-informative Tokens**:
   - Removed stop words (e.g., "the," "is") and punctuation to focus on the meaningful content.
   - Also excluded tokens with specific Part-of-Speech (POS) tags, such as pronouns, determiners, particles, symbols, and spaces. This further reduces irrelevant information, focusing on nouns, verbs, and adjectives that are likely to convey the main themes.

5. **Conversion to Lowercase**:
   - Converted all tokens to lowercase to avoid case-sensitive differences in words, ensuring consistency across the text.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.3 Build a topic model using sklearn's LatentDirichletAllocation
rubric={points}

**Your tasks:**

1. Build LDA models on the preprocessed data using using [sklearn's `LatentDirichletAllocation`](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.LatentDirichletAllocation.html) and random state 42. Experiment with a few values for the number of topics (`n_components`). Pick a reasonable number for the number of topics and briefly justify your choice.

<div class="alert alert-warning">

Solution_2_3
    
</div>

_Points:_ 4

We selected 8 topics because it provides distinct, interpretable themes that align well with the categories in the dataset (e.g., sports, religion, politics, technology) without excessive overlap or over-specificity.

In [47]:
from sklearn.decomposition import LatentDirichletAllocation
from sklearn.feature_extraction.text import CountVectorizer

# Convert preprocessed text data to a document-term matrix
vectorizer = CountVectorizer(max_df=0.95, min_df=2) 
dtm = vectorizer.fit_transform(df["text_pp"])

# Experiment with different values for the number of topics
n_topics_options = [5, 8, 10, 15] 
lda_models = {}
for n_topics in n_topics_options:
    lda = LatentDirichletAllocation(n_components=n_topics, random_state=42)
    lda.fit(dtm)
    lda_models[n_topics] = lda

# Display the top words for each topic for different values of n_components
def display_topics(model, feature_names, num_top_words=10):
    for topic_idx, topic in enumerate(model.components_):
        top_words = [feature_names[i] for i in topic.argsort()[:-num_top_words - 1:-1]]
        print(f"Topic #{topic_idx + 1}: {' '.join(top_words)}")

print("LDA topics for different n_components:")
for n_topics, lda_model in lda_models.items():
    print(f"\nNumber of topics: {n_topics}")
    display_topics(lda_model, vectorizer.get_feature_names_out())

LDA topics for different n_components:

Number of topics: 5
Topic #1: num edu period new pub jpeg play la vs game
Topic #2: god people think know believe jesus num say thing time
Topic #3: file program use window image num entry widget server display
Topic #4: num gun people state israel right law jews weapon government
Topic #5: num year say go game think people know come team

Number of topics: 8
Topic #1: num period la new pts vs edu pp st play
Topic #2: god jesus num know find book church question bible time
Topic #3: num file window program use entry widget server set line
Topic #4: num gun law file firearm weapon state control crime bill
Topic #5: game year team num go say think good know play
Topic #6: num armenian israel turkish jews people armenians israeli government jewish
Topic #7: image file format color graphic jpeg software available ftp package
Topic #8: people think god know believe say like thing right life

Number of topics: 10
Topic #1: num period la pts vs new play

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.4 Exploring word topic association
rubric={points}

**Your tasks:**
1. For the number of topics you picked in the previous exercise, show top 10 words for each of your topics and suggest labels for each of the topics (similar to how we came up with labels "health and nutrition", "fashion", and "machine learning" in the toy example we saw in class). 

> If your topics do not make much sense, you might have to go back to preprocessing in Exercise 2.1, improve it, and train your LDA model again. 

<div class="alert alert-warning">

Solution_2_4
    
</div>

_Points:_ 5

In [48]:
vectorizer = CountVectorizer(max_df=0.95, min_df=2, stop_words='english')
dtm = vectorizer.fit_transform(df['text_pp'])

lda = LatentDirichletAllocation(n_components=8, random_state=42)
lda.fit(dtm)

def display_topics(model, feature_names, n_top_words):
    topics = []
    for topic_idx, topic in enumerate(model.components_):
        top_words = [feature_names[i] for i in topic.argsort()[:-n_top_words - 1:-1]]
        topics.append(f"Topic #{topic_idx+1}: {' '.join(top_words)}")
    return topics

feature_names = vectorizer.get_feature_names_out()
topics = display_topics(lda, feature_names, 10)
for topic in topics:
    print(topic)

Topic #1: armenian jews people turkish num armenians say year genocide government
Topic #2: israel israeli right num state people greek arab attack country
Topic #3: num file program image use window available include entry server
Topic #4: num period la play vs pts pp power goal new
Topic #5: num say god people know come jesus think time tell
Topic #6: year point think islam career morris kk read big day
Topic #7: gun god people think believe know question law thing claim
Topic #8: game team good player think play year like know time


## Topic #1 Events in Armenian History
**Why**: These words talk about what happened to Armenian people in the past

## Topic #2 Middle East News
**Why**: Words are about events and conflicts in Israel and nearby countries

## Topic #3 Computer Programs
**Why**: Words are about using and making computer software

## Topic #4 Game Scores
**Why**: Words are about points and goals in sports games

## Topic #5 God and Faith
**Why**: Words are about religious beliefs and Jesus

## Topic #6 Islam and Life
**Why**: Words are about Islam and everyday experiences

## Topic #7 Gun Rules 
**Why**: Words are about guns and rules about them

## Topic #8 Sports Teams
**Why**: Words are about how teams and players do in sports

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.5 Exploring document topic association
rubric={points}

**Your tasks:**
1. Show the document topic assignment of the first five documents from `df`. 
2. Comment on the document topic assignment of the model. 

<div class="alert alert-warning">

Solution_2_5
    
</div>

_Points:_ 5

In [55]:
# Get document-topic distributions
doc_topics = lda.transform(dtm)

for i in range(5):
    print(f"Document {i + 1}:")
    # Get probabilities higher than 1%
    topic_probs = [(j+1, prob) for j, prob in enumerate(doc_topics[i]) if prob > 0.01]
    # Sort by probability in descending order
    topic_probs.sort(key=lambda x: x[1], reverse=True)
    
    # Display topic probabilities
    for topic_num, prob in topic_probs:
        print(f"Topic {topic_num}: {prob:.3f}")
    print(f"Dominant Topic: {doc_topics[i].argmax() + 1}")
    print()

Document 1:
Topic 7: 0.823
Topic 4: 0.104
Topic 2: 0.062
Dominant Topic: 7

Document 2:
Topic 6: 0.931
Topic 4: 0.059
Dominant Topic: 6

Document 3:
Topic 7: 0.482
Topic 3: 0.384
Topic 5: 0.123
Dominant Topic: 7

Document 4:
Topic 8: 0.506
Topic 5: 0.473
Dominant Topic: 8

Document 5:
Topic 1: 0.639
Topic 2: 0.284
Topic 6: 0.062
Dominant Topic: 1



Document 1:
This document is strongly focused on Gun Rules (Topic 7) with 82.3% probability, with a minor connection to Game Scores (Topic 4). Since Gun Rules is so dominant, this document is likely a discussion about gun legislation or gun control policies.

Document 2:
This document shows an extremely strong alignment with Islam and Life (Topic 6) at 93.1%. The high probability for a single topic suggests this is a highly focused discussion about Islamic practices or daily life in relation to Islam.

Document 3:
This document shows a mixed distribution between Gun Rules (Topic 7, 48.2%) and Computer Programs (Topic 3, 38.4%), with some religious content (God and Faith, Topic 5, 12.3%). This combination might suggest a discussion that bridges multiple topics, perhaps about technology in relation to gun rights or regulations.

Document 4:
This document is almost evenly split between Sports Teams (Topic 8, 50.6%) and God and Faith (Topic 5, 47.3%). This is an interesting combination, which might indicate a discussion about religion in sports or perhaps a sports team with religious beliefs.

Document 5:
This document is primarily about Armenian History (Topic 1, 63.9%) with significant coverage of Middle East News (Topic 2, 28.4%). This combination makes sense as these topics are geographically related, suggesting a discussion about Armenian history in the context of broader Middle Eastern events.

<!-- END QUESTION -->

<br><br><br><br>

<!-- BEGIN QUESTION -->

## Exercise 3: Short answer questions 
<hr>

rubric={points}

1. Briefly explain how content-based filtering works in the context of recommender systems. 
2. Discuss at least two negative consequences of recommender systems.
3. What is transfer learning in natural language processing? Briefly explain.     

<div class="alert alert-warning">

Solution_3
    
</div>

_Points:_ 6

1. The content-based filtering works by looking at what a user likes and finding similar items for them. For example, if you watch love movies with Leonardo DiCaprio, it will recommend other love movies or more Leonardo DiCaprio movies. 

2. 
- Filter Bubbles: People get stuck seeing only things they already agree with or like, making them less open to new ideas or different viewpoints.
- Privacy Issues: These systems collect lots of personal data about what we like, watch, and buy, which can be misused or leaked.

3. Transfer learning is a machine learning technique in which knowledge gained through one task or dataset is used to improve model performance on another related task and/or different dataset. It's like teaching someone a new language when they already know a similar one. The system first learns basic language skills from a large amount of text (like learning English), and then uses that knowledge to do specific tasks better (like understanding movie reviews). It's more efficient because it doesn't have to learn everything from scratch.

<!-- END QUESTION -->

<br><br><br><br>

**Before submitting your assignment, please make sure you have followed all the instructions in the Submission instructions section at the top.** 

![](img/eva-well-done.png)